book/data_viz/plotting_data.nim

```diff
@@ -7,18 +7,521 @@ nbUseNimibook
+nbText: """
+This looks a bit better. At least we have something that sort of resembles our
+input data! But what's that wide gray bar from 0 to roughly 3?
+Our data frame covers the `x` range *twice*. At the intersection from the first,
+- at index 24 - our "Energy" column is `10.0`. From there it jumps back to
+0.0 on the next bin. This leads to a full bin that accidentally covers the full
+range from 0 to 10. Let's check that assumption by printing values between index
+24 and 26 from our data frame:
+"""
+nbCodeInBlock:
+  echo df[24 .. 26]
+nbText: """
+As we can see index 24 with a value of `2.746` is used for a bin with bin width 10.
+
+To get closer to the plot we want, we will perform classification by a discrete
+variable. Let's `color` by the "Type" column:
+"""
+nbCodeInBlock:
+  ggplot(df, aes("Energy", "Counts", color = "Type")) +
+    geom_histogram(stat = "identity") +
+    ggsave("images/multi_layer_histogram_2.png")
+nbImage("images/multi_layer_histogram_2.png")
+nbText: """
+Ohh, interesting. See how we have an automatic legend based on the
+two classes found in column "Type".
+
+Well, our plot wants to run out of our graph, which is probably what
+you were focusing on, isn't it? (Note: this is also a bug. For some reason
+stacking isn't properly handled right now for the data range
+calculation. Issue #99). Let's ignore data running out of the plot for
+now, because we can't see anything anyway.
+
+But didn't we say we want to have classification by `color`? For bars
+`color` refers to the *outline* of a bar. We need to add a `fill` to get the
+bars into a fully colored object.
+"""
+nbCodeInBlock:
+  ggplot(df, aes("Energy", "Counts", color = "Type", fill = "Type")) +
+    geom_histogram(stat = "identity") +
+    ggsave("images/multi_layer_histogram_3.png")
+nbImage("images/multi_layer_histogram_3.png")
+nbText: """
+Aha! Now we begin to see why the data is running out of our plot. Apparently
+both classes are being *stacked* on top of one another. This is the default
+behavior for classified histograms so that all the data is visible. Without
+transparency we would hide data otherwise.
+
+Aside from the bug that causes the data to run out of the plot (we could
+manually set a range using `ylim`). To change this behavior to the one we want
+we will apply `position = "identity"`:
+"""
+nbCodeInBlock:
+  ggplot(df, aes("Energy", "Counts", color = "Type", fill = "Type")) +
+    geom_histogram(stat = "identity", position = "identity") +
+    ggsave("images/multi_layer_histogram_4.png")
+nbImage("images/multi_layer_histogram_4.png")
+nbText: """
+This is already looking somewhat reasonable, barring the fact that we now
+have the exact problem stacking is supposed to fix. One histogram overlaps
+the other. We can solve that by applying 50% alpha channel:
+"""
+nbCodeInBlock:
+  ggplot(df, aes("Energy", "Counts", color = "Type", fill = "Type")) +
+    geom_histogram(stat = "identity", position = "identity", alpha = some(0.5)) +
```

Comment:
Might be worth mentioning `some(0.5)` is related to `Option` so people know what to search for further info on it?

Reply:
Maybe a good idea, yes.

Reply:
Idk what level of Nim-knowledge we are aiming for really but `Option` feels like something even an experienced Nim developer could have missed if they simply never had the use for something like that. Is the reason for using an `Option` that different choices of the other parameters need different defaults?

Reply:
Yes, using `Option` makes the API much more sane. Otherwise it becomes a mess of "well, if the value is 0 it probably means the user has not specified a size". I've forgotten the worst offenders that made me choose this path right now. For some cases it's fine, but in others it becomes more tricky.

In addition internally it makes it immediately obvious if something is user given (`some`) or not. 
It's kind of a pity that it's not possible to have auto converters from "if argument is given by user, wrap in `Option`". That would be the ideal.

It would be easily achievable with a macro of course. `aes` is already a macro. I'm not sure if *every* `geom_*` proc should be a macro though. Kind of worried about the implications of that.

So for now...

Reply:
Haha yeah sounds like a mess for sure 😅👍 

I mean, there are `converters` but we might not want to export a `float -> Option[float]` converter that could affect other parts of their code 🤷 

If one can avoid making things into macros one should keep to procs, updating macros when on get a new idea isn't fun...

Reply:
> I mean, there are converters but we might not want to export a float -> Option[float] converter that could affect other parts of their code shrug

Exactly that I want to avoid. I need to play around if it's enough to have such a converter locally in a module without exporting them. If so that might actually be possible that way (but probably not, because the conversion must probably be done *before* the overload is chosen, which means it must happen in the calling scope).

Reply:
Yeah, that sounds farfetched to work :/